### PR TITLE
Feat/bounded streams && refactor streaming pallet

### PR DESCRIPTION
**Pull request by rjman-ljm**

### CODE REVIEW

File: pallets/streaming/src/lib.rs [hunk 1]
```diff
@@ -381,65 +396,46 @@ impl<T: Config> Pallet<T> {
...
-                .ok_or(DispatchError::Arithmetic(ArithmeticError::Underflow))
-        } else {
-            stream
-                .stop_time
-                .checked_sub(stream.start_time)
-                .ok_or(DispatchError::Arithmetic(ArithmeticError::Underflow))
-        }
+    pub fn update_finished_stream_library(account: &AccountOf<T>) -> DispatchResult {
+        let checked_pop =
+            |regisry: &mut Option<BoundedVec<StreamId, T::MaxStreamsCount>>| -> DispatchResult {
```
Comment (i5possible) [inline pallets/streaming/src/lib.rs:401]:
```suggestion
            |registry: &mut Option<BoundedVec<StreamId, T::MaxStreamsCount>>| -> DispatchResult {
```

File: pallets/streaming/src/lib.rs [hunk 2]
```diff
@@ -381,65 +396,46 @@ impl<T: Config> Pallet<T> {
...
-        } else {
-            Ok(Zero::zero())
-        }
+    pub fn try_push_stream_library(
+        account: &AccountOf<T>,
+        kind: StreamKind,
+        stream_id: StreamId,
+    ) -> DispatchResult {
+        let checked_push =
+            |regisry: &mut Option<BoundedVec<StreamId, T::MaxStreamsCount>>| -> DispatchResult {
```
Comment (i5possible) [inline pallets/streaming/src/lib.rs:430]:
```suggestion
            |registry: &mut Option<BoundedVec<StreamId, T::MaxStreamsCount>>| -> DispatchResult {
```

File: pallets/streaming/src/lib.rs [hunk 3]
```diff
@@ -381,65 +396,46 @@ impl<T: Config> Pallet<T> {
...
-            stream
-                .stop_time
-                .checked_sub(stream.start_time)
-                .ok_or(DispatchError::Arithmetic(ArithmeticError::Underflow))
-        }
+    pub fn update_finished_stream_library(account: &AccountOf<T>) -> DispatchResult {
+        let checked_pop =
+            |regisry: &mut Option<BoundedVec<StreamId, T::MaxStreamsCount>>| -> DispatchResult {
+                let mut r = regisry.take().unwrap_or_default();
+                r.as_mut().sort_unstable_by(|a, b| b.cmp(a));
```
Comment (i5possible) [inline pallets/streaming/src/lib.rs:403]:
```suggestion
                r.as_mut().reserve();
```

Comment (i5possible) [reply pallets/streaming/src/lib.rs:403]: Would reverse work? Or it's changed permanently?

**Review by GopherJ — APPROVED** [review]

File: pallets/streaming/src/lib.rs [hunk 4]
```diff
@@ -381,65 +396,46 @@ impl<T: Config> Pallet<T> {
...
+
+                let len = r.len() as u32;
+                match len {
+                    _x if len >= T::MaxFinisehdStreamsCount::get() => {
+                        if let Some(index) = r.pop() {
```
Comment (i5possible) [inline pallets/streaming/src/lib.rs:408]: Remove the sort_unstable_by and use r.first() ?

File: pallets/streaming/src/mock.rs [hunk 5]
```diff
@@ -146,12 +146,14 @@ impl pallet_currency_adapter::Config for Test {
 parameter_types! {
     pub const StreamPalletId: PalletId = PalletId(*b"par/strm");
     pub const MaxStreamsCount: u32 = 128;
+    pub const MaxFinisehdStreamsCount: u32 = 2;
```
Comment (i5possible) [inline pallets/streaming/src/mock.rs:149]: It's the finished streams count or it's canceled or both?

Comment (i5possible) [inline pallets/streaming/src/mock.rs:149]:
```suggestion
    pub const MaxFinishedStreamsCount: u32 = 2;
```

**Review by i5possible — APPROVED** [review]
Looks good to me expect for some typo.

File: pallets/streaming/src/lib.rs (hunk 4 above)
Comment (rjman-ljm) [reply pallets/streaming/src/lib.rs:408]: image

Comment (rjman-ljm) [reply pallets/streaming/src/lib.rs:408]: 🤔

File: pallets/streaming/src/mock.rs (hunk 5 above)
Comment (rjman-ljm) [reply pallets/streaming/src/mock.rs:149]: StreamLibrary::<T>::try_mutate(account, &StreamKind::Finish, checked_pop)?;

Comment (rjman-ljm) [reply pallets/streaming/src/mock.rs:149]: It's the finished streams count, include `Cancelled` and `Completed`

**Comment by rjman-ljm** [comment]
image

**Pull request merged**